### DISCUSSION TITLE
Service Interactions

### DISCUSSION THREAD

**Discussion by georgestevens99** [adamreed90/dapr-samples#5] [Ideas]
Our first goal is to demonstrate how people can use the new way of ServiceInvocation.  Thus we can present an example of the most basic simple demo of how Service Invocation is used in a typical microservices system, i.e.  the typical Client-->WebAPI-->gRPCService call chain that is often found in microservices. This can be demonstrated in 2 scenarios as follows:

SayHello:
Console.SayHello()--http-->WebAPIAhttp.SayHello()--Dapr.ServiceInvocation-->ServiceAgRpc.SayHello()  
The ServiceAgRPC.SayHello() method returns "hello" which is bubbled back to the Console via the rpc call sequence returns.

Echo:
Console.Echo(AStringToEcho)--http-->WebAPIAhttp.Echo(stringReceivedFromCaller)--Dapr.ServiceInvocation-->ServiceAgRpc.Echo(stringReceivedFromCaller)  
The ServiceAgRPC.Echo() method returns the stringReceivedFromCaller which is bubbled back to the Console via the rpc call sequence returns.

Once we get that working, then there are other interactions that we can demonstrate as well if you wish, either in the current project or in a follow on.  Plus the same applies to State Management and PubSub, not to mention Actors, etc, etc.

To me the most important thing is to make simple, easy to understand examples of situations typically found in real world programming.  The other kind of stuff really isn't nearly so helpful IMO.

Finally, to be realistic, the Console app should be able to process simple commands that allow the execution of one, or multiple commands in a loop.  The multiple command feature is easy to implement and really useful to see how the system performs under load.

For example, the command:  echo  A-string-to-echo 100        

This will cause 100 echo calls to the WebAPI.  I have used 10,000 just for fun in my POCs (and they never crashed, even once!).  The syntax is commandName commandArg numberOfIterations.  This is really simple to parse.

So there you have my vision of what Phase 1 of this project can be like.  I hope you are onboard with this vision, keeping in mind once implemented it provides a basic starter solution for a lot more samples!

Please let me know what you think.  Is this a good feature set to get started with, quickly making a high quality sample and building the base for more?  Suggestions?  Ideas?  @paulgyro

**Comment by iamadamreed** [comment]
I definitely agree on all points, I think at our base we should cover the basics, Console, HttpAPI, gRPC Api, Actor, PubSub, and State Stores. 
 I hoped to also add a Blazor WASM using JWT example, I guess we can call that and others extra credit! :D

**Reply by paulgyro** [reply]
Ok it's settled in my mind; this is really going to end up being a reference architecture.  We will certainly need to progress towards that but between the three of us I think we'll have the bandwidth to really make a killer reference and/or series of samples
